### DISCUSSION TITLE
Automation of AI downloads from bananas

### DISCUSSION THREAD

**Discussion by gregdek** [OpenTTD/OpenTTD#8418] [Q&A]
I'm trying to set up a script that downloads random AIs from Bananas to a dedicated OpenTTD server, runs games with them, and records the results. Why? Because it seems like fun. :)

What's the right approach?

Approach 1: figure out how to connect to the admin port and use the Bananas downloader in-game. I presume this is the best approach because I presume that the in-game client does dependency resolution and downloads the necessary libraries as well as the AI. I've been playing with various tools, but I can't quite figure out how they work. I can get the clients from libottdadmin2 to connect, but it's not clear to me how to extend them. 

Approach 2: just download all the AI libraries directly from the Bananas site -- there aren't that many -- and install them, and then build a list of AIs and download them directly as well. Downside is that I'd be building, in essence, a static set of content. Which would be fine, I suppose.

Any hints? I think I'd prefer Approach 1, because I think learning the admin server API will be useful in a lot of ways -- but there's just not enough to sink my teeth into. Even a couple of clearer example scripts for libottdadmin2 would be extremely helpful.

Thanks for any tips. Once I understand things better, I'm also happy to help write docs and whatnot.

**Comment by TrueBrain** [comment]
We used to do this in the beginning of the AI framework. We would have "tournament" servers, where AIs would battle each other. We would rate them, and publish the results. That was so much fun :D
Sadly, as more and more AIs showed up, this became impractical to do on the resources available, so we abandoned the project. But given CPUs got a lost faster in 10 years, this is an awesome idea :D

Approach 2 is not going to work, as you cannot download content from the website directly; you have to do it from in-game.

What we used to do, was modify the server slightly, but that was before the admin port exists :P But I do not think that the system is designed that well for such automation; I do not think you can list a category and download all or a subset of AIs, honestly.

What you could do, is start the game yourself, select all the AIs, download them (all dependencies will be downloaded correctly), and use that to run your games with. This means you can list on disk which AIs are available, and you can look inside the `tar` to find the name of the AI. This you can use again to generate an `openttd.cfg` which sets the AI-names that should be started for each company.

It will not be trivial, this challenge you set, but it is doable. It depends a bit on what your goal is exactly, but here are some other tips:

- by starting OpenTTD with `-g -vnull:ticks=30000` you can limit that a new game only runs for 30,000 ticks.
- you can use `-g` too to give a savegame. This means you can generate a savegame and let all AIs battle on the same map.
- you can use `-d ai=1` or something to get some more debug logging what AIs are doing. Could be useful for debug hunting.
- we used to modify the server slightly to on-exit of the game print out the company value (and some other stats) about the AI companies. This was used to rate how each one was doing.
- you can make a screenshot ever N ticks, so you can see how AIs progressed. I have never done this myself, but I know others have; a bit of searching on https://tt-forums.net should result in some examples of that. I believe they set the camera to a given location, and run the screenshot command every N ticks. This is really neat to add for such games. I believe a few bigger servers used to do this .. and I believe it can be done via the admin port. But again, never did it myself, so a bit of hearsay :)

Hmm .. think that is about it, I think. If you have any questions, feel free to poke me on [Discord](https://discord.gg/hKzMGUx) or on [IRC](https://webchat.oftc.net/?channels=openttd); there will be more people in both places able to help you out too :)

Good luck with your venture; it is a fun one I can promise you :D Don't forget to let us know the results when you got them, I am sure many people are interested. Best place for that would be the [forums](https://tt-forums.net) or [Discord](https://discord.gg/hKzMGUx) :D

**Comment by gregdek** [comment]
Thanks for the hints.

I basically downloaded a bunch of AIs (all of them, essentially) on my local machine, tarred them up, and dropped them on to my server, just to get me unblocked -- but now after having played with the scripts, and having had some success with parameterized generation of on_dedicated.scr and game_start.scr scripts, I'm wondering if I can download AIs as part of game_start.scr.  

As far as scoring, I was thinking I'd try to write a simple scorekeeper AI or gamescript, which could spit out logs of arbitrary complexity -- I was thinking it might be fun to try to feed various stats into an ELK stack to get realtime logging of games, but that's a ways down the line yet. :) Not sure if that approach will work, but if it does, that would allow me to easily externalize any victory conditions. 

Thanks for all of your work on this awesome game.
